lib/bech32/src/Codec/Binary/Bech32/Internal.hs

```diff
@@ -169,12 +180,41 @@ dataPartToText (DataPart t) = t
 -- > dataPartToWords (dataPartFromWords w) == w
 --
 dataPartFromWords :: [Word5] -> DataPart
-dataPartFromWords = DataPart . T.pack . fmap (word5ToChar Arr.!)
+dataPartFromWords = DataPart . T.pack . fmap dataCharFromWord
 
 -- | Convert a 'DataPart' into words.
 --
 dataPartToWords :: DataPart -> [Word5]
-dataPartToWords = mapMaybe charToWord5 . T.unpack . dataPartToText
+dataPartToWords = mapMaybe dataCharToWord . T.unpack . dataPartToText
+
+-- | Returns true iff. the specified character is permitted to appear within
+--   the data part of a Bech32 string.
+--   See here for more details: https://git.io/fj8FS
+dataCharIsValid :: Char -> Bool
+dataCharIsValid = (`Map.member` dataCharToWordMap)
+
+-- | A list of all characters that are permitted to appear within the data part
+--   of a Bech32 string. See here for more details: https://git.io/fj8FS
+dataCharList :: String
+dataCharList = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+-- | If the specified character is permitted to appear within the data part
+--   of a Bech32 string, this function returns that character's corresponding
+--   'Word5' value. If the specified character is not permitted, or if the
+--   specified character is upper-case, returns 'Nothing'.
+dataCharToWord :: Char -> Maybe Word5
+dataCharToWord = (`Map.lookup` dataCharToWordMap)
+
+dataCharToWordMap :: Map Char Word5
+dataCharToWordMap = Map.fromList $ dataCharList `zip` (Word5 <$> [0 .. 31])
+
+-- | Maps the specified 'Word5' onto a character that is permitted to appear
+--   within the data part of a Bech32 string.
+dataCharFromWord :: Word5 -> Char
+dataCharFromWord = (dataCharFromWordArray Arr.!)
+
+dataCharFromWordArray :: Array Word5 Char
+dataCharFromWordArray = Arr.listArray (Word5 0, Word5 31) dataCharList
```

Comment:
Why not use a `Map Word5 Char` here too in the end :thinking: ?

Reply:
> Why not use a `Map Word5 Char` here too in the end

Good question.

The mapping from `Word5` to `Char` covers **all** values of `Word5`. So the type we need is actually `Word5 -> Char`. There should be no valid value of `Word5` that doesn't map to some `Char`.

Using `Data.Map.lookup` gives us a type of `Word5 -> Maybe Char`. So when performing a lookup, we'd have to pattern match on the result, and then call `error` in the case that the lookup fails.

What do you think?

Reply:
Okay, different needs so different data-structures. I didn't quite grasp that at first but it makes sense indeed to use `Array` here.